src/opnsense/www/api.php

```diff
@@ -1,56 +1,37 @@
 <?php
 
-try {
-    /**
-     * Read the configuration
-     */
-    $config = include __DIR__ . "/../mvc/app/config/config.php";
-
-    /**
-     * Read auto-loader
-     */
-    include __DIR__ . "/../mvc/app/config/loader.php";
-
-    /**
-     * Read services
-     */
-    include __DIR__ . "/../mvc/app/config/services_api.php";
-
-    /**
-     * Handle the request
-     */
-    $application = new \Phalcon\Mvc\Application($di);
-
-    echo $application->handle($_SERVER['REQUEST_URI'])->getContent();
-} catch (\Error | \Exception $e) {
-    $unbuffered = strpos($e->getMessage(), 'ob_end_clean') !== false;
-
-    if (!is_a($e, 'OPNsense\Base\UserException') && !$unbuffered) {
-        error_log($e);
-    }
-
-    if ($unbuffered) {
-        /**
-         * if buffer has been deleted already, we must assume the implementation
-         * has already sent headers and we cannot send them again.
-         */
-        return;
+function error_output($http_code, $e,  $user_message)
+{
+    $response = [];
+    if (OPNsense\Core\Config::getInstance()->object()->system->deployment != 'development'){
+        $response['errorMessage'] = $user_message;
+    } else {
+        $response['errorMessage'] = $e->getMessage();
+        $response['errorTrace'] = $e->getTraceAsString();
     }
-
-    $response = [
-        'errorMessage' => $e->getMessage(),
-        'errorTrace' => $e->getTraceAsString(),
-    ];
-
     if (method_exists($e, 'getTitle')) {
         $response['errorTitle'] = $e->getTitle();
-    } else {
-        $response['errorTitle'] = gettext('An API exception occurred');
-        $response['errorMessage'] = $e->getFile() . ':' . $e->getLine() . ': ' . $response['errorMessage'];
     }
-
-    header('HTTP', true, 500);
+    header('HTTP', true, $http_code);
     header("Content-Type: application/json;charset=utf-8");
+    echo htmlspecialchars(json_encode($response), ENT_NOQUOTES);
+}
+
 
-    echo json_encode($response, JSON_UNESCAPED_SLASHES);
+try {
+    $config = include __DIR__ . "/../mvc/app/config/config.php";
+    include __DIR__ . "/../mvc/app/config/loader.php";
+
+    $router = new OPNsense\Mvc\Router('/api/', 'Api');
+    $response = $router->routeRequest($_SERVER['REQUEST_URI']);
+    if (!$response->isSent()) {
```

Comment:
Feels like sending a rocket to the moon. Who could send() before? Mid-term maybe just preventing send after send() is called and not checking isSent() seems more straightforward.

Reply:
Unfortunately there are more quirks in phalcon, in practice the controller already has called `send()`, which means this shouldn't be needed, but to keep backwards compatibility I've left the old behavior here (send if nobody did)